transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+			int qos;
+		} subscribe;
+		struct {
+			char *topic;
+			int qos;
+		} publish;
+	} admin;
+} janus_mqtt_context;
+
+/* Transport client methods */
+void janus_mqtt_client_connection_lost(void *context, char *cause);
+int janus_mqtt_client_message_arrived(void *context, char *topicName, int topicLen, MQTTAsync_message *message);
+void janus_mqtt_client_delivery_complete(void *context, MQTTAsync_token token);
+int janus_mqtt_client_connect(janus_mqtt_context *ctx);
+void janus_mqtt_client_connect_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_connect_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_reconnect(janus_mqtt_context *ctx);
+void janus_mqtt_client_reconnect_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_reconnect_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_disconnect(janus_mqtt_context *ctx);
+void janus_mqtt_client_disconnect_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_disconnect_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_subscribe(janus_mqtt_context *ctx);
+void janus_mqtt_client_subscribe_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_subscribe_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_admin_subscribe(janus_mqtt_context *ctx);
+void janus_mqtt_client_admin_subscribe_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_admin_subscribe_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_publish_janus_message(janus_mqtt_context *ctx, char *payload);
+void janus_mqtt_client_publish_janus_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_publish_janus_failure(void *context, MQTTAsync_failureData *response);
+int janus_mqtt_client_publish_admin_message(janus_mqtt_context *ctx, char *payload);
+void janus_mqtt_client_publish_admin_success(void *context, MQTTAsync_successData *response);
+void janus_mqtt_client_publish_admin_failure(void *context, MQTTAsync_failureData *response);
+void janus_mqtt_client_destroy_context(janus_mqtt_context *ctx);
+
+/* We only handle a single client */
+static janus_mqtt_context *context_ = NULL;
+
+int janus_mqtt_init(janus_transport_callbacks *callback, const char *config_path) {
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* Initializing context */
+	janus_mqtt_context *ctx = g_malloc0(sizeof(struct janus_mqtt_context));
+	ctx->gateway = callback;
+	context_ = ctx;
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_MQTT_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL)
+		janus_config_print(config);
+
+	/* Handle configuration */
```

Comment:
All this should happen in the `if(config != NULL)` check. It's not that all unlikely to load a plugin without a configuration file available (e.g., a user forgetting to do a `make configs`).